lib/block_view_validator.go

```diff
@@ -0,0 +1,1493 @@
+	// Validate ValidatorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Validate Domains.
+	if len(metadata.Domains) < 1 {
+		return RuleErrorValidatorNoDomains
+	}
+	if len(metadata.Domains) > MaxValidatorNumDomains {
+		return RuleErrorValidatorTooManyDomains
+	}
+	var domainStrings []string
+	for _, domain := range metadata.Domains {
+		_, err := url.ParseRequestURI(string(domain))
+		if err != nil {
+			return fmt.Errorf("%s: %v", RuleErrorValidatorInvalidDomain, domain)
+		}
+		domainStrings = append(domainStrings, string(domain))
+	}
+	if len(NewSet(domainStrings).ToSlice()) != len(domainStrings) {
+		return RuleErrorValidatorDuplicateDomains
+	}
+
+	// TODO: In subsequent PR, validate VotingPublicKey, VotingPublicKeySignature, and VotingSignatureBlockHeight.
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnregisterAsValidatorMetadata(transactorPublicKey []byte, metadata *UnregisterAsValidatorMetadata) error {
+	// Validate ValidatorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Validate ValidatorEntry exists.
+	validatorEntry, err := bav.GetValidatorByPKID(transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "IsValidUnregisterAsValidatorMetadata: ")
+	}
+	if validatorEntry == nil {
+		return RuleErrorValidatorNotFound
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) GetValidatorByPKID(pkid *PKID) (*ValidatorEntry, error) {
+	// First check the UtxoView.
+
+	// There can be multiple ValidatorEntries for a given PKID in the UtxoView since the ValidatorMapKey
+	// contains ValidatorPKID, TotalStakeAmountNanos, and RegisteredAtBlockHeight. We need to loop through
+	// all the ValidatorEntries and find the one matching the given PKID that is !isDeleted. There should
+	// ever only be zero or one such matching ValidatorEntries. If the only matching ValidatorEntries are
+	// all isDeleted then we shouldn't check the database as the corresponding rows in the database will
+	// be deleted once the UtxoView is flushed.
+	isDeleted := false
+
+	for _, validatorEntry := range bav.ValidatorMapKeyToValidatorEntry {
+		if validatorEntry == nil {
+			// This should never happen but is a sanity check.
+			continue
+		}
+		if !validatorEntry.ValidatorPKID.Eq(pkid) {
+			continue
+		}
+		if validatorEntry.isDeleted {
+			isDeleted = true
+			continue
+		}
+		// If we get to this point, we found a matching
+		// !isDeleted ValidatorEntry for the given PKID.
+		return validatorEntry, nil
+	}
+
+	if isDeleted {
+		// If we get to this point, we found one or more matching ValidatorEntries
+		// for the given PKID, but they were all isDeleted. We do not want to check
+		// the database but instead just return nil, no ValidatorEntry found.
+		return nil, nil
+	}
```

Comment:
hmm is it possible that we have a delete entry on the view but a valid one in the DB? Should we get all the validator entries from the DB (there should only be one) and merge it with the view and then do this logic? I'm considering the case where we had a disconnect - I *think* it's okay, but I need to think a little more on it.

Reply:
No. I don't think it's possible. There's only one (or zero) valid !isDeleted ValidatorEntry for the same PKID that can exist at a time. There can be multiple ValidatorEntries in the UtxoView since we store TotalStakeAmountNanos and RegisteredAtBlockHeight in the UtxoView ValidatorMapKey. But only one should ever be !isDeleted at a time for each PKID.

If someone stakes, then there will be two ValidatorEntries corresponding to the same PKID in the UtxoView: the old ValidatorEntry with the original TotalStakeAmountNanos and isDeleted=true. And the new ValidatorEntry with the updated TotalStakeAmountNanos and isDeleted=false. If someone unregisters and then reregisters within the same block, there will similarly be two ValidatorEntries in the UtxoView corresponding to the same PKID: the old ValidatorEntry with the original RegisteredAtBlockHeight and isDeleted=true. And the new ValidatorEntry with the updated RegisteredAtBlockHeight and isDeleted=false.

If the only ValidatorEntry for this PKID found in the view has isDeleted=true, then once the view is flushed there won't be any ValidatorEntry rows corresponding to this PKID in the database. There's only ever one row in the database corresponding to each ValidatorEntry PKID. It's only if there aren't any isDeleted or !isDeleted ValidatorEntries in the UtxoView for this PKID that we should check the db. LMK if this makes it clearer or more confusing. :upside_down_face:

Reply:
Basically the same conclusion I arrived at, but explained much more thoroughly. Thanks!